### PR TITLE
Aws/bugs

### PR DESCRIPTION
**Pull request by jonathangold**
This patch fixes 2 bugs in longpollWatch()

https://github.com/purpleidea/mgmt/commit/be50e3f05da54ea71636def9eeea4b5e6000f459 The first is that after 10-15 minutes, the waiters return with "ResourceNotReady: exceeded wait
attempts" which we need to ignore and continue.

https://github.com/purpleidea/mgmt/commit/152b9cbec30b271ab47ccec342e41f7b6f09a3f9 The second takes care of the concurrent waiters for stopped and terminated as discussed.

### CODE REVIEW

File: resources/aws_ec2.go [hunk 1]
```diff
@@ -394,6 +394,9 @@ func (obj *AwsEc2Res) longpollWatch() error {
 	}()
 	obj.wg.Add(1)
 	go func() {
+		// shorten some errors we need throughout
+		errCancelled := request.CanceledErrorCode
```
Comment (purpleidea) [inline resources/aws_ec2.go]: amazingly, this is actually a constant! https://godoc.org/github.com/aws/aws-sdk-go/aws/request#pkg-constants

let's not re-define it, let's just use it directly.

File: resources/aws_ec2.go [hunk 2]
```diff
@@ -466,23 +466,47 @@ func (obj *AwsEc2Res) longpollWatch() error {
 						},
 					}
 					log.Printf("%s: Watching: %s", obj, *diOutput.Reservations[0].Instances[0].InstanceId)
-					if err := obj.client.WaitUntilInstanceStoppedWithContext(ctx, waitInput); err != nil {
-						if aerr, ok := err.(awserr.Error); ok {
-							if aerr.Code() == errCancelled {
-								log.Printf("%s: Request cancelled", obj)
+					// Wait for instance to stop or
+					// terminate concurrently.
+					wg := sync.WaitGroup{}
```
Comment (purpleidea) [inline resources/aws_ec2.go]: Cool!

**Comment by jonathangold** [comment]
@purpleidea Agreed. I think that in the long-run, longpollWatch and CheckApply would both benefit from being broken down into helper funcs/methods, much like snsWatch is. At the very least, it would eliminate a bunch of channel selects and make for easier reading.
For now I'll just encapsulate the goroutines in their own method.

File: resources/aws_ec2.go [hunk 3]
```diff
@@ -1014,6 +994,65 @@ func (obj *AwsEc2Res) prependName() string {
 	return AwsPrefix + obj.GetName()
 }
 
+// waitForStopAndTerminate waits for the instance to stop and waits for it to
+// terminate. If either waiter returns, the instance state is checked, and an
+// awsEc2Event is returned.
+func (obj *AwsEc2Res) waitForStopAndTerminate(ctx context.Context, waitInput *ec2.DescribeInstancesInput) (awsEc2Event, error) {
```
Comment (purpleidea) [inline resources/aws_ec2.go]: do we want to name this the absurd `waitUntilInstanceStoppedOrTerminatedWithContext` to match AWS naming conventions?

Maybe it could also have the same signature? as WaitUntilInstanceStoppedWithContext (so return error)

IOW we pretend it's identical to `WaitUntilInstanceStoppedWithContext` or `WaitUntilInstanceTerminatedWithContext` except it waits for either.

File: resources/aws_ec2.go [hunk 4]
```diff
@@ -1014,6 +994,65 @@ func (obj *AwsEc2Res) prependName() string {
...
+	go func() {
+		defer wg.Done()
+		defer wgCancel()
+		if err := obj.client.WaitUntilInstanceTerminatedWithContext(wgCtx, waitInput); err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				if aerr.Code() == request.CanceledErrorCode || aerr.Code() == request.WaiterResourceNotReadyErrorCode {
+					return
+				}
+			}
+			select {
```
Comment (purpleidea) [inline resources/aws_ec2.go]: This shouldn't be part of this function if you agree with the above comment about matching the AWS API.

File: resources/aws_ec2.go [hunk 5]
```diff
@@ -1014,6 +994,65 @@ func (obj *AwsEc2Res) prependName() string {
...
+			case obj.awsChan <- &chanStruct{
+				err: errwrap.Wrapf(err, "unknown error waiting for instance to terminate"),
+			}:
+			case <-obj.closeChan:
+			}
+		}
+	}()
+	wg.Wait()
+	// Check the instance state, and return the appropriate event.
```
Comment (purpleidea) [inline resources/aws_ec2.go]: As well, this should also be outside of this function if you agree with my initial API comment. Of course if this type of logic happens often enough throughout the code, it's fine to put it in a helper function!

File: resources/aws_ec2.go [hunk 6]
```diff
@@ -1014,6 +994,78 @@ func (obj *AwsEc2Res) prependName() string {
...
+		defer wgCancel()
+		if err := obj.client.WaitUntilInstanceStoppedWithContext(wgCtx, waitInput); err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				if aerr.Code() == request.CanceledErrorCode || aerr.Code() == request.WaiterResourceNotReadyErrorCode {
+					return
+				}
+			}
+			select {
+			case errChan <- errwrap.Wrapf(err, "unknown error waiting for instance to stop"):
+			case <-obj.closeChan:
```
Comment (purpleidea) [inline resources/aws_ec2.go]: `obj.*` shouldn't be in here-- remember this is like a new AWS method. Actually it probably shouldn't even be part of `*AwsEc2Res`, but if you want to use a closeChan pattern, you need to pick a private var inside this function.

External to this function, which could get closed by the context, that context closing might trigger an internal close via the internal closeChan! :)

File: resources/aws_ec2.go [hunk 7]
```diff
@@ -1014,6 +994,78 @@ func (obj *AwsEc2Res) prependName() string {
...
+					return
+				}
+			}
+			select {
+			case errChan <- errwrap.Wrapf(err, "unknown error waiting for instance to terminate"):
+			case <-obj.closeChan:
+			}
+		}
+	}()
+	select {
```
Comment (purpleidea) [inline resources/aws_ec2.go]: This pattern isn't correct-- the two go routines will start running, and then assuming something isn't immediately available on errChan, default will trigger mean we *will* return nil. The only catch is we'll only see that happen after the two go-routines close and the waitGroup lets us exit.

So you've actually just masked any chance of error.

**Comment by jonathangold** [comment]
Sorry. I pushed this code prematurely. I'll work on it tomorrow and make sure the logic makes sense. :P

**Comment by jonathangold** [comment]
Sorry. I pushed this code prematurely. I'll work on it tomorrow and make sure the logic makes sense. :P

File: resources/aws_ec2.go [hunk 8]
```diff
@@ -1014,6 +994,91 @@ func (obj *AwsEc2Res) prependName() string {
...
+	defer close(errChan)
+	wg := sync.WaitGroup{}
+	defer wg.Wait()
+	closeChan := make(chan struct{})
+	wgCtx, wgCancel := context.WithCancel(ctx)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case <-wgCtx.Done():
```
Comment (jonathangold) [inline resources/aws_ec2.go]: @purpleidea Not %100 sure if this is the right approach to closing closeChan. Is there a chance that this could block? IOW, if there's an error, is it possible that wgCtx won't get cancelled so closeChan doesn't get closed?

File: resources/aws_ec2.go [hunk 9]
```diff
@@ -1014,6 +994,91 @@ func (obj *AwsEc2Res) prependName() string {
...
+			}
+		}
+	}()
+	select {
+	case err, ok := <-errChan:
+		if !ok {
+			return fmt.Errorf("error reading from errChan")
+		}
+		return err
+	case <-wgCtx.Done():
```
Comment (purpleidea) [inline resources/aws_ec2.go]: what does the AWS api return if the context gets cancelled? :P

File: resources/aws_ec2.go [hunk 10]
```diff
@@ -1014,6 +994,91 @@ func (obj *AwsEc2Res) prependName() string {
...
+			select {
+			case errChan <- errwrap.Wrapf(err, "unknown error waiting for instance to terminate"):
+			case <-closeChan:
+			}
+		}
+	}()
+	select {
+	case err, ok := <-errChan:
+		if !ok {
+			return fmt.Errorf("error reading from errChan")
```
Comment (purpleidea) [inline resources/aws_ec2.go]: this is not what this means, and not what we'd want to return if errChan closes-- we'd want to use that channel closed signal as a "return nil" (without error!)

File: resources/aws_ec2.go [hunk 11]
```diff
@@ -1014,6 +994,91 @@ func (obj *AwsEc2Res) prependName() string {
...
+// or the other.
+func waitUntilInstanceStoppedOrTerminatedWithContext(ctx context.Context, waitInput *ec2.DescribeInstancesInput, c *ec2.EC2) error {
+	errChan := make(chan error)
+	defer close(errChan)
+	wg := sync.WaitGroup{}
+	defer wg.Wait()
+	closeChan := make(chan struct{})
+	wgCtx, wgCancel := context.WithCancel(ctx)
+	wg.Add(1)
+	go func() {
```
Comment (purpleidea) [inline resources/aws_ec2.go]: You can make this work without having a third goroutine AFAICT.

File: resources/aws_ec2.go (hunk 10 above)
Comment (purpleidea) [inline resources/aws_ec2.go]: this is not what this means-- rephrase your error message please and explain why you want this scenario to be an error. I'm okay with this, but you have to tell me why.

File: resources/aws_ec2.go [hunk 12]
```diff
@@ -1014,6 +994,91 @@ func (obj *AwsEc2Res) prependName() string {
...
+			select {
+			case errChan <- errwrap.Wrapf(err, "unknown error waiting for instance to stop"):
+			case <-closeChan:
+			}
+		}
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer wgCancel()
```
Comment (purpleidea) [inline resources/aws_ec2.go]: what's this? you shouldn't do that here.
and you didn't close(closeChan) -- which is probably what you want to do here. you can cheat by doing this via a https://golang.org/pkg/sync/#Once if you want.

File: resources/aws_ec2.go [hunk 13]
```diff
@@ -641,11 +618,14 @@ func (obj *AwsEc2Res) longpollWatch() error {
 			}
 		case msg, ok := <-obj.awsChan:
 			if !ok {
-				return *exit
+				return fmt.Errorf("channel closed unexpectedly")
```
Comment (jonathangold) [inline resources/aws_ec2.go]: If you want me to put this in another patch I don't mind. It came up when I was testing and I hadn't noticed.

Comment (purpleidea) [reply resources/aws_ec2.go]: Should this be an error or nil?

File: resources/aws_ec2.go [hunk 14]
```diff
@@ -1014,6 +994,101 @@ func (obj *AwsEc2Res) prependName() string {
...
+		}
+	}
+	return awsEc2EventNone, nil
+}
+
+// waitUntilInstanceStoppedOrTerminatedWithContext combines the two waiters
+// required to trigger events if a running instance is stopped or terminated.
+// This function is needed, because the AWS api only provides waiters for one
+// or the other.
+func waitUntilInstanceStoppedOrTerminatedWithContext(ctx context.Context, waitInput *ec2.DescribeInstancesInput, c *ec2.EC2) error {
```
Comment (purpleidea) [inline resources/aws_ec2.go]: Nice implementation here! ;)

File: resources/aws_ec2.go [hunk 15]
```diff
@@ -466,45 +466,22 @@ func (obj *AwsEc2Res) longpollWatch() error {
...
-							err: errwrap.Wrapf(err, "unknown error waiting for instance to stop"),
-						}:
-						case <-obj.closeChan:
-						}
-						return
-					}
-					stateOutput, err := obj.client.DescribeInstances(diInput)
+					// Wait for instance to stop or
+					// terminate concurrently.
+					event, err := longpollRunningWaiter(ctx, waitInput, obj.client)
```
Comment (purpleidea) [inline resources/aws_ec2.go]: RE: https://github.com/jonathangold/mgmt/blob/de5caffce49a6d69156c8b1853670b4d18930ca0/resources/aws_ec2.go#L471-L485

are there any returns needed like the other ones in the other select's?

**Comment by purpleidea** [comment]
Merged thanks!
